codice/ddf
# DDF-05061 Refactor WFS sources to not use dynamically-generated attributes

#### What does this PR do?
DDF's WFS sources dynamically generate attributes of the form 'ext.FeatureTypeName', where each attribute corresponds to a WFS feature property. These attributes are exposed throughout the system, including Intrigue, which causes confusion for users because the attributes seem to appear out of nowhere.

DDF already provides the ability to map metacard attributes to WFS feature properties and those mappings should be used instead of the dynamically-generated attributes, which were only intended to be used internally.

This PR completely removes the dynamically-generated attributes. Users will now have to create mappings between metacard attributes and feature properties in order to populate metacards with data from their WFS source.

Note: the WFS 1.0 and WFS 2.0 sources do not yet support query mapping between metacard attributes and feature properties. Query mapping will be added for those sources in follow-up PRs.

#### Who is reviewing it? 
@glenhein 
@dcruver 
@leo-sakh 
@gjvera 
@bennuttle 

#### Select relevant component teams: 
@codice/io 
@codice/ogc 

#### Ask 2 committers to review/merge the PR and tag them here.
@jlcsmith
@troymohl

#### How should this be tested?
1. In the Admin UI, open the Spatial app.
2. Click `Metacard to WFS Feature Map`.
3. For `Feature Type`, enter `{http://stategeothermaldata.org/uri-gin/aasg/xmlschema/welllog/0.8}AZWellLogs`.
4. For `Metacard Title to WFS Feature Property Mapping`, enter one of the element names from http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer?request=DescribeFeatureType&service=WFS&version=1.1.0 that has a type of `xsd:string` such as `DisplayName` or `Notes`.
a. Do the same for `Resource URI` and try to map it to a element.
b. Map `Resource Size` to an element of type `xsd:double` such as `DrillerTotalDepth_ft`.
5. For `Metacard Created Date to WFS Feature Property Mapping`, enter one of the element names from http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer?request=DescribeFeatureType&service=WFS&version=1.1.0 that has a type of `xsd:dateTime` such as `StatusDate`, `SpudDate`, or `DateTimeLogRun`.
a. Do the same for `Modified`, `Effective`, and `Expiration` and try to map each to a different element.
6. For `Metacard Geography to WFS Feature Property Mapping`, enter `Shape`.
7. Click `Save changes`.
8. Create a WFS 1.1.0 source with the URL http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer
a. Disable the CN check and allow redirects. Leave everything else as-is.
9. Open Intrigue and run queries against the WFS source using the metacard attributes you've mapped (e.g. `Title`, `Resource URI`, `Resource Size`).
a. Check http://services.azgs.az.gov/ArcGIS/services/aasggeothermal/AZWellLogs/MapServer/WFSServer?request=GetFeature&service=WFS&version=1.1.0&typeName=AZWellLogs to see what the features look like and to find values you can query for.
b. For example, if you've mapped `Title` to `DisplayName` then you should be able to search for `Title` `CONTAINS` `*Gamma*` or `Title` `CONTAINS` `*Lithologic*`.
c. This source does not appear to support temporal queries so queries on `Created`/`Modified`/`Effective`/`Expiration` will not work.
10. For each query, inspect the results in Intrigue using the table or inspector view and verify:
a. The results match your query.
b. The metacard attributes you mapped have been correctly populated with data from the WFS feature properties. `Title`, `Resource URI`, `Location`, `Created`, `Modified`, etc. should be filled out with the data from the corresponding WFS feature property.
c. You can run `log:set info org.apache.cxf` and inspect the `FeatureCollection` response in the logs to see if the metacard attributes match the corresponding feature properties of your query results.
11. Verify there are no `ext.AZWellLogs.*` attributes in the table view or the inspector.

#### What are the relevant tickets?
Fixes: #5061 

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
